Update chrome-php to version 1.11 and Add PDF Generation Enhancements

tsmilan opened the pull request:
This merge request includes the following updates and additions:
- Updated chrome-php to version 1.11
- Added support for custom window size and user agent in PDF generation
- Introduced wait_for_js_condition method to handle waiting for a JavaScript condition with a timeout
- Integrated FPDI library via composer for PDF merging functionality
- Added support for converting and combining multiple Moodle URLs into a single PDF
- Preserved sesskey in the URL to prevent session expiration
- Updated PHP version requirement in README.md
- Bumped the version and updated supported Moodle versions

tsmilan committed:
Add support for custom window size and user agent in PDF generation

tsmilan committed:
Add method to wait for JavaScript condition with timeout

tsmilan committed:
Add support for converting and combining multiple Moodle URLs into a single PDF

tsmilan committed:
Preserve sesskey in the URL to prevent session expiration

tsmilan committed:
Update PHP version requirement in README.md

tsmilan committed:
Version bump and update supported Moodle versions

cameron1729 commented on classes/converter.php:
```diff
@@ -17,6 +17,10 @@
 namespace tool_pdfpages;
 
 use moodle_url;
+use setasign\Fpdi\TcpdfFpdi;
```
What's this `setasign` thing? Could we use `pdf` from **lib/pdflib.php** instead?

cameron1729 replied:
Thanks for discussing this one with me @tsmilan. I always thought `assignfeedback_editpdf` was using core's PDF wrapper to do the stitching, but after checking with you, I see it bundles this `TcpdfFpdi` thing in the feedback plugin itself and uses that to join pages

It would be nice if there was something in core we could use to join the pages, but as the `TcpdfFpdi` thing from assign is part of the feedback plugin, it doesn't seem appropriate to use that here. So I think what you've done in bundling it in to this plugin itself is appropriate.

cameron1729 commented on classes/converter.php:
```diff
@@ -96,6 +100,119 @@ final public function convert_moodle_url_to_pdf(moodle_url $url, string $filenam
+     *
+     * @return \stored_file the stored file created during conversion.
+     */
+    final public function convert_moodle_urls_to_pdf(array $urls, string $filename = '', array $options = [],
```
Should we add some validation to ensure `$urls` is actually a list of `moodle_url`s ?

tsmilan replied:
Hi Cam, I pushed a fixup commit here https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/17b313ae0e886ef0be9153ce2c62676724bc2547

cameron1729 commented on classes/converter_chromium.php:
```diff
@@ -117,6 +121,51 @@ protected function generate_pdf_content(moodle_url $proxyurl, string $filename =
+     *                      Defaults to 30000ms (30 seconds).
+     * @throws \Exception If MathJax does not finish rendering within the specified timeout.
+     */
+    protected function wait_for_mathjax_condition(Page $page, int $timeout = 30000): void {
```
It feels a bit odd to have this function that's specific to mathjax. Is there anyway it could be implemented more generally? Like `wait_for_javascript` or something like that?

tsmilan replied:
I pushed a fixup commit here. Updated it to be `wait_for_js_condition`  https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/aaac411fe1e289664771b5bb50e7b6789085daa8 

PS: I also pushed some commits to for improvements

cameron1729 replied:
Hey @tsmilan it seems like this might be causing issues with the **test.php** script. When I try this: http://localhost/mymoodle/admin/tool/pdfpages/test.php?url=http://localhost/mymoodle/course/view.php?id=3 I'm getting an error:

Exception - tool_pdfpages\converter_chromium::wait_for_js_condition(): Argument #3 ($jsconditionparams) must be of type array, null given, called in [dirroot]/admin/tool/pdfpages/classes/converter_chromium.php on line 115

tsmilan replied:
Hi Cam, I pushed a fix here https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/d4f67bd153ca6ac7e9696307decf64f2235ec002

You might need to disable shm if you're using a docker container as mentioned here https://github.com/chrome-php/chrome/issues/446#issuecomment-1307304445 to avoid the operation timed out error after 30s
```
    protected function generate_pdf_content(moodle_url $proxyurl, string $filename = '', array $options = [],
                                            string $cookiename = '', string $cookievalue = '',
                                            array $renderoptions = []): string {
        try {
            $browseroptions = [
                'headless' => true,
                'noSandbox' => true,
                'customFlags' => [
                    '--disable-dev-shm-usage',
                ],
            ];
```

cameron1729 replied:
Test page is working for me now :)

tsmilan committed:
Add Chromium-specific rendering options

tsmilan committed:
Update capability check to support system and module context levels

tsmilan committed:
Update ci.yml

tsmilan committed:
Update version.php

tsmilan committed:
Update thirdpartylibs.xml

- Remove unused packages
- Add Monolog to third party libs to avoid Moodle CI check issues

tsmilan committed:
Add vendor/setasign to thirdpartylibs.xml

cameron1729 commented on README.md:
```diff
@@ -51,7 +58,7 @@ git clone git@github.com:catalyst/moodle-tool_pdfpages.git <moodledir>/admin/too
 Use of the converter requires programmatic access, there in no frontend associated with this plugin, so you need to develop another module, or add this plugin to the dependencies of an existing Moodle plugin.
 
-> Only users with the system level capability `tool/pdfpages:generatepdf` can conducted conversions, as this is required to create the single use access key for the headless browser session.
+> Only users with the system or module level capability `tool/pdfpages:generatepdf` can conducted conversions, as this is required to create the single use access key for the headless browser session.
```
Minor typo. "conducted" -> "conduct".

tsmilan replied:
I pushed a fix https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/0122d7e62055e3788637789489b78cdc26bb0888

cameron1729 commented on classes/converter.php:
```diff
@@ -89,6 +101,87 @@ final public function convert_moodle_url_to_pdf(moodle_url $url, string $filenam
+            bool $printpagenumbers = false): \stored_file {
+        global $USER, $PAGE;
+
+        $allurlsarevalid = empty(array_filter($urls, fn($url) => !$url instanceof moodle_url));
```
Let's make it `fn($url): bool =>...`

tsmilan replied:
Added fixup commit https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/40f56da8d76a7e2fc0dbc4c47087e57b70bee9ab

tsmilan committed:
Fetch page context via a Closure

Fetch the page context by using a Closure to avoid calling the magic method "magic_get_context",
as it triggers a debugging coding problem for pages that don't have a context.

tsmilan committed:
Fix typo in README.md

cameron1729 commented on README.md:
```diff
@@ -51,7 +58,7 @@ git clone git@github.com:catalyst/moodle-tool_pdfpages.git <moodledir>/admin/too
 Use of the converter requires programmatic access, there in no frontend associated with this plugin, so you need to develop another module, or add this plugin to the dependencies of an existing Moodle plugin.
 
-> Only users with the system level capability `tool/pdfpages:generatepdf` can conducted conversions, as this is required to create the single use access key for the headless browser session.
+> Only users with the system or module level capability `tool/pdfpages:generatepdf` can conduct conversions, as this is required to create the single use access key for the headless browser session.
```
I think this isn't quite accurate, is it? It's not about system or module context, it's about whether the user's role in a specific context has the capability.

Like I could assign a user a role in course context that has the capability, and it would still work, right?

It also mentions "this is required to create the single use access key". Do you know what's meant by that? Is it the session key?

tsmilan replied:
Updated it to say "in a specific context" xD https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/7325fffcf13d151213ca65d758ad063c7125fbb7

The access key is what they are using to login and allow access of target page (I read from the docblock). When we visit the URL of the page we want to print, instead of logging in manually via username/password or SAML, we are using access keys to validate and login the user. Session key is different, that's what we get after we login.

cameron1729 commented on README.md:
```diff
@@ -78,7 +85,7 @@ __Note__: if you didn't specify a filename when converting, you can obtain the f
 This will utilise the configured converter on the server side to carry out the conversion, creating the converted file in the Moodle file system and then serve up the PDF to the browser.
 
-In order to access this page, the logged in Moodle user needs to be an Admin or have a role with the capability `tool/pdfpages:generatepdf` at the system level.
+In order to access this page, the logged in Moodle user needs to be an Admin or have a role with the capability `tool/pdfpages:generatepdf` at the system or module context level.
```
As with the other part of the readme, I think this may need to say something like "have a role with the capability `blah` in the relevant context".

tsmilan replied:
Updated https://github.com/catalyst/moodle-tool_pdfpages/pull/40/commits/7325fffcf13d151213ca65d758ad063c7125fbb7 xD

cameron1729 commented on classes/converter_chromium.php:
```diff
@@ -69,17 +70,30 @@ class converter_chromium extends converter {
     protected function generate_pdf_content(moodle_url $proxyurl, string $filename = '', array $options = [],
-                                              string $cookiename = '', string $cookievalue = ''): string {
+                                            string $cookiename = '', string $cookievalue = '',
+                                            array $renderoptions = []): string {
```
Can we re-use the existing `$options` param instead of adding a new one? The docblock kinda hints that this is what it was originally for:

`array $options any additional options to pass to converter, valid options vary with converter`

tsmilan replied:
The `$options` is specific to Chrome-php options. We're using it as `$page->pdf($options);`  I haven't tried passing unsupported `$options` there. Here's their list of valid options https://github.com/chrome-php/chrome?tab=readme-ov-file#available-options 

Edit: actually that's not the options, it's this one https://github.com/chrome-php/chrome/blob/1.11/src/PageUtils/PagePdf.php#L27

Should I try to put our custom options there?

tsmilan replied:
I have updated it to use `$options` sorry I rebased :sweat_smile:

tsmilan committed:
Update chrome-php to version 1.11

tsmilan committed:
Add FPDI library via composer for PDF merging functionality

cameron1729 commented:
Thanks @tsmilan - this one looks good to merge.
